### PR TITLE
Configure the option --reconcileWebhookConfig based on DNS certificate provisioning for Sidecar Injector

### PR DESCRIPTION
**Pull request by lei-tang**
Configure the option --reconcileWebhookConfig of Sidecar Injector based on DNS certificate provisioning: (the approach is this PR is same as that of https://github.com/istio/istio/pull/16118).

- If DNS certificates are provisioned, --enable-reconcileWebhookConfiguration is set as false.
- Otherwise, --enable-reconcileWebhookConfiguration is set as true.

Please provide a description for what this PR is for: https://github.com/istio/istio/issues/17061

Design documents:
[1]https://docs.google.com/document/d/1VHyBre8943USOx_YEVtA18KfEoGTmLT1iBHESO5bzcA/edit#heading=h.qex63c29z2to
[2] https://docs.google.com/document/d/1v8BxI07u-mby5f5rCruwF7odSXgb9G8-C9W5hQtSIAg/edit#heading=h.xw1gqgyqs5b

The matching PR on istio-installer repo: https://github.com/istio/installer/pull/428

And to help us figure out who should review this PR, please
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[X ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/values.yaml
Comment (howardjohn) [inline install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/values.yaml]: can you expand on this - what does manage mean? what is webhook configuration? why should I set this?

I think I know, but these docs are user facing and the average user may not

**Comment by costinm** [comment]
Please no PR in installer that doesn't link to an equivalent change in the
new installer/operator !!

**Comment by costinm** [comment]
And the naming of the new setting is horrible - how would a user know what
it means ?

Besides - I would like to see a doc, and have it discussed in Env WG WRT
all the webhook management (in particular since it
affects multi-cluster and install options).

Current mechanism is broken - and with a-la-carte and operator we must be
able to separate the installation of the component
( galley, injector ) from the enablement of the webhook, as well as have a
way to turn it off easily.
Best approach seems to be istioctl and new operator.

**Comment by costinm** [comment]
To avoid confusion: I am strongly in favor of not having galley or injector
manage their webhook config, and doing this in
a separate step/component.  But we need to take into account several
goals/requirements - as well as operator plans, that's
why I want a design first and review/signoff by all interested parties.

I wrote down some comments on the cluster-wide settings separation -
https://github.com/istio/installer/wiki/WIP:-Istio-Cluster-config
The webhooks are cluster-wide, and in order to support canaries and
a-la-carte and better security for deployment will
need to be separated from the deployment one way or another.

File: install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/values.yaml
Comment (lei-tang) [reply install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/values.yaml]: Done: the explanation of this parameter has been expanded.

**Comment by lei-tang** [comment]
@costinm 
- The description of this PR has included "The matching PR on istio-installer repo: istio/installer#310." 
- The naming of the parameter has been changed from "enableManageWebhookConfig" to "enableSidecarInjectorToManageWebhookConfig". If you have a suggestion for the naming of the parameter, please let us know.
- The design document for Istio webhook management is at: https://docs.google.com/document/d/1VHyBre8943USOx_YEVtA18KfEoGTmLT1iBHESO5bzcA. The design has been presented at Istio design review meetings, reviewed, and approved. If you like to have another meeting to discuss the design, please let me know and I will schedule a meeting with related stakeholders.

File: install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/templates/deployment.yaml
Comment (morvencao) [inline install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/templates/deployment.yaml]: Can we set a default value here by adding `| default <true_or_false>`?

Comment (lei-tang) [reply install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/templates/deployment.yaml]: The default value has been set in install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/values.yaml. Moreover, if changing the line into " - --enableSidecarInjectorToManageWebhookConfig={{ .Values.enableSidecarInjectorToManageWebhookConfig | default true }}", the helm set command does not set the value of enableSidecarInjectorToManageWebhookConfig based on the value in the helm set command, e.g., the command "helm template --name=istio --namespace=istio-system --set sidecarInjectorWebhook.enableSidecarInjectorToManageWebhookConfig=false install/kubernetes/helm/istio > istio-test.yaml" will not set enableSidecarInjectorToManageWebhookConfig to false.

Comment (morvencao) [reply install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/templates/deployment.yaml]: Why not? 
I did't have a chance to try `helm template --name=istio --namespace=istio-system --set sidecarInjectorWebhook.enableSidecarInjectorToManageWebhookConfig=false install/kubernetes/helm/istio`
I thought the parameter in `--set` in `helm install` has the highest priority.

Comment (lei-tang) [reply install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/templates/deployment.yaml]: According to https://github.com/helm/helm/blob/master/docs/chart_template_guide/values_files.md: "--set has a higher precedence than the default values.yaml file". But I am unable to find a helm document describing the precedence relationship between the helm default function and the helm set command. If you can find such document, please let us know.

**Comment by rlenglet** [comment]
I met offline with @costinm and we agreed on the following plan of action:

1. We need a way to disable webhook configuration in Citadel and Injection Webhook, and such flags are the right way to implement feature flags in the `istio/istio` installer.
1. The flags can be added into `istio/istio` installer. **So we will submit this PR.**
1. The flags will not be added into the new `istio/installer`, since webhook configuration will be disabled by default in Citadel and Injection Webhook.
1. We need to test the logic and combinations of flags, especially upgrade from 1.3. **This is tracked in https://github.com/istio/istio/issues/16860.**

**Review by rlenglet — APPROVED** [review]

**Comment by rlenglet** [comment]
@istio/wg-environments-maintainers could you please review this PR?

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
